statistics/statistics.go

```diff
@@ -21,329 +21,143 @@ package statistics
 import (
 	"errors"
 	"fmt"
-	"math"
-	"sort"
-	"time"
+	"strings"
 
-	log "github.com/Sirupsen/logrus"
 	"github.com/intelsdi-x/snap-plugin-lib-go/v1/plugin"
-	"github.com/montanaflynn/stats"
 )
 
+type Plugin struct {
+	buffer map[string]*dataBuffer
+}
+
 const (
 	Name    = "statistics"
-	Version = 2
+	Version = 3
 )
 
-type Plugin struct {
-	buffer        map[string][]interface{}
-	bufferMaxSize int
-	bufferCurSize int
-	bufferIndex   int
-}
+var (
+	statList = []string{"count", "mean", "median", "standard_deviation", "variance", "95%_ile", "99%_ile", "2%_ile", "9%_ile", "25%_ile", "75%_ile", "91%_ile", "98%_ile", "minimum", "maximum", "range", "mode", "kurtosis", "skewness", "sum", "trimean", "first_quartile", "third_quartile", "quartile_range"}
+)
 
 // New() returns a new instance of this
 func New() *Plugin {
-	buffer := make(map[string][]interface{})
-	p := &Plugin{buffer: buffer,
-		bufferMaxSize: 100,
-		bufferCurSize: 0,
-		bufferIndex:   0}
+	buffer := make(map[string]*dataBuffer)
+	p := &Plugin{buffer: buffer}
 	return p
 }
 
-// calculateStats calaculates the descriptive statistics for buff
-func (p *Plugin) calculateStats(buff interface{}, startTime time.Time, stopTime time.Time, namespace string, unit string) ([]plugin.Metric, error) {
-	var result []plugin.Metric
-	var buffer []float64
-	var err error
-	var val float64
-	var modeVal []float64
-	tags := map[string]string{
-		"startTime": startTime.String(),
-		"stopTime":  stopTime.String(),
-	}
-	time := time.Now()
+// GetConfigPolicy returns the config policy
+func (p *Plugin) GetConfigPolicy() (plugin.ConfigPolicy, error) {
+	policy := plugin.NewConfigPolicy()
 
-	//Need to change so it ranges over the current size of the buffer and not the capacity
-	for _, val := range buff.([]interface{}) {
-		switch v := val.(type) {
-		default:
-			st := fmt.Sprintf("Unknown data received in calculateStats(): Type %T", v)
-			return nil, errors.New(st)
-		case int:
-			buffer = append(buffer, float64(val.(int)))
-		case int32:
-			buffer = append(buffer, float64(val.(int32)))
-		case int64:
-			buffer = append(buffer, float64(val.(int64)))
-		case float64:
-			buffer = append(buffer, val.(float64))
-		case float32:
-			buffer = append(buffer, float64(val.(float32)))
-		case uint64:
-			buffer = append(buffer, float64(val.(uint64)))
-		case uint32:
-			buffer = append(buffer, float64(val.(uint32)))
-		}
-	}
+	policy.AddNewIntRule([]string{""}, "slidingWindowLength", false, plugin.SetDefaultInt(100), plugin.SetMinInt(1))
+	policy.AddNewIntRule([]string{""}, "slidingFactor", false, plugin.SetDefaultInt(1), plugin.SetMinInt(1))
+	policy.AddNewStringRule([]string{""}, "statistics", false, plugin.SetDefaultString(strings.Join(statList, ",")))
```

Comment:
If nothing is passed from config, all statistics are returned. This is might be more user friendly, especially if people try the plugin for the first time, or if we want to showcase with simple task manifest.